File: 3406-find-all-possible-stable-binary-arrays-i/find-all-possible-stable-binary-arrays-i.cpp
class Solution {
public:
    static const int MOD = 1e9 + 7;

    int limit;
    int dp[201][201][2];

    int solve(int zero, int one, int last) {

        if (zero == 0 && one == 0)
            return 1;

        if (dp[zero][one][last] != -1)
            return dp[zero][one][last];

        long long ans = 0;

        if (last == 0) {

            for (int take = 1; take <= limit && take <= one; take++) {
                ans += solve(zero, one - take, 1);
                ans %= MOD;
            }

        } else {

            for (int take = 1; take <= limit && take <= zero; take++) {
                ans += solve(zero - take, one, 0);
                ans %= MOD;
            }
        }

        return dp[zero][one][last] = ans;
    }

    int numberOfStableArrays(int zero, int one, int lim) {

        limit = lim;

        memset(dp, -1, sizeof(dp));

        long long ans = 0;

       
        for (int take = 1; take <= limit && take <= zero; take++)
            ans = (ans + solve(zero - take, one, 0)) % MOD;

       
        for (int take = 1; take <= limit && take <= one; take++)
            ans = (ans + solve(zero, one - take, 1)) % MOD;

        return ans;
    }
};

File: 3406-find-all-possible-stable-binary-arrays-i/find-all-possible-stable-binary-arrays-i.java
import java.util.*;

class Solution {

    int MOD = 1000000007;
    int limit;
    Long[][][] dp;

    public int numberOfStableArrays(int zero, int one, int limit) {
        this.limit = limit;
        dp = new Long[zero + 1][one + 1][2];

        long ans = 0;

        for (int i = 1; i <= Math.min(limit, zero); i++) {
            ans = (ans + solve(zero - i, one, 0)) % MOD;
        }

        for (int i = 1; i <= Math.min(limit, one); i++) {
            ans = (ans + solve(zero, one - i, 1)) % MOD;
        }

        return (int) ans;
    }

    private long solve(int zero, int one, int last) {
        if (zero == 0 && one == 0) {
            return 1;
        }

        if (dp[zero][one][last] != null) {
            return dp[zero][one][last];
        }

        long ways = 0;

        if (last == 0) {
            for (int i = 1; i <= Math.min(limit, one); i++) {
                ways = (ways + solve(zero, one - i, 1)) % MOD;
            }
        } else {
            for (int i = 1; i <= Math.min(limit, zero); i++) {
                ways = (ways + solve(zero - i, one, 0)) % MOD;
            }
        }

        dp[zero][one][last] = ways;
        return ways;
    }
}